acousticbrainz: Write tags to file

JOJ0 opened the pull request:
## Description

- Fixes #3928
- Adds the ability to write to actual file tags using the acousticbrainz plugin.
- ~Requires https://github.com/beetbox/mediafile/pull/59~
- Adds an info log statement when AcousticBrainz data could not be fetched due to simply not having an MBID.

## To Do

- [x] Documentation. (If you've add a new command-line flag, for example, find the appropriate page under `docs/` to describe it.)
- [x] Changelog. (Add an entry to `docs/changelog.rst` near the top of the document.)
- ~[ ] Tests. (Encouraged but not strictly required.)~

JOJ0 commented:
Here's an example of the written metadata (provided the requirede metadata changes):
```
TXXX=ab:hi:danceability:danceable=0.000000000000
TXXX=ab:hi:gender=female
TXXX=ab:hi:genre_rosamerica=rhy
TXXX=ab:hi:ismir04_rhythm=ChaChaCha
TXXX=ab:hi:mood_acoustic:acoustic=0.081802986562
TXXX=ab:hi:mood_aggressive:aggressive=0.000000000000
TXXX=ab:hi:mood_electronic:electronic=0.979390621185
TXXX=ab:hi:mood_happy:happy=0.085078120232
TXXX=ab:hi:mood_party:party=0.000015778420
TXXX=ab:hi:mood_relaxed:relaxed=0.808817088604
TXXX=ab:hi:mood_sad:sad=0.109234951437
TXXX=ab:hi:moods_mirex=Cluster5
TXXX=ab:hi:timbre=dark
TXXX=ab:hi:tonal_atonal:tonal=0.002889123978
TXXX=ab:hi:voice_instrumental=instrumental
TXXX=ab:lo:average_loudness=0.887623310089
TXXX=ab:lo:tonal:chords_changes_rate=0.082755811512
TXXX=ab:lo:tonal:chords_key=A#
TXXX=ab:lo:tonal:chords_number_rate=0.002038320526
TXXX=ab:lo:tonal:chords_scale=minor
TXXX=ab:lo:tonal:key_strength=0.654201686382
```

JOJ0 commented:
From my point of view there only is one open question in this PR: The new mediafile fields use 12 float_places, this plugin is configured to use only 6 if I understand correctly: https://github.com/JOJ0/beets/blob/acousticbrainz_write_tags/beetsplug/acousticbrainz.py#L114-L135

In the post above we see the 12 float_places are taken care of when writing the tags to a file. Also in `beet edit ..` we see the full number with 12 places. Also in the database it is like that.

Only in `beet info` it shows 6 places after the comma.

I would like to understand when the configured datatypes of the plugin itself (link above to L114-L135) are used and when they are not. Not only interesting for this PR but also for future contributions. Thanks for helping me read the code :-)

sampsyo commented:
Hi! I'm actually a little confused about why this change is necessary. I believe that the `try_write` call should _already_ write all the file's tags—including the AcousticBrainz ones. If you have a MediaFile with those tags supported, I thought there wouldn't need to be any change to beets—have you given that a shot?

JOJ0 commented:
Yes, my first tries were assuming that nothing should be required in the plugin, but then discovered, that as documented here: https://github.com/beetbox/beets/blob/19e4f41a72ade8e69dfb43b38a6569985ecefd79/beets/library.py#L733-L734

and respected here: https://github.com/beetbox/beets/blob/19e4f41a72ade8e69dfb43b38a6569985ecefd79/beets/library.py#L757-L758

a field to be written is required to either be in this list:
https://github.com/beetbox/beets/blob/19e4f41a72ade8e69dfb43b38a6569985ecefd79/beets/library.py#L460

or passed as the tags kwarg to the try_write/write method as I did.

So the alternative is to add to the mentioned _fields list, but that didn't feel correct to me since we only want those fields when the plugin is in use.

sampsyo commented:
Ah, silly me! Thanks for the reminder about `_media_fields`, which does indeed exclude arbitrary flexible attributes from being written automatically.

With that reminder, it occurs to me that the other way to do this is for the plugin to add things directly to `_media_fields`, as is the case with the plugin API `add_media_field`:
https://github.com/beetbox/beets/blob/018396ccf74a35e56ab346ad390476a0e347307a/beets/plugins.py#L205

And this also got me thinking… feel free to reject this thought, but given the impending doom of AcousticBrainz as a project, maybe the right solution here is for the plugin itself to use `add_media_field` (rather than changing MediaFile itself)? That is, changing MediaFile made more sense when we thought AB tags were here to stay and would only get more popular and supported by more and more tools. But now that this no longer seems to be the case, and the tags are going to be more of a niche thing, maybe we needn't muck with the broader tagging abstraction and are better off adding the tags directly in this beets plugin. What do you think?

JOJ0 commented:
Hi @sampsyo, yes I hear you and the `add_media_field` option seems like the right way to go (BTW: Amazing how flexible beets is and how many possible ways to solve problems are existing already! Again, amazing project, love it! Haha!)

But,....I had the following in mind for quite some time already: There is a plugin existing that uses the Essentia extractors directly to analyze local files, not relying on AcousticBrainz but _almost_ using the exact same field names:  https://github.com/adamjakab/BeetsPluginXtractor#xtractor-beets-plugin

My plan was to contribute the tagging to actual file tags over there as well, so it would make sense to have the changes in MediaFile directly and require minimal changes in that plugin.

At that time I thought that actually it is good that no new tag names are invented for this idea and the AB: tags are perfectly suitable for that. And still we could say they wouldn't be the worst idea....

But, .....now that won't speed up my intention of finally gettings this feature merged ;-).....actually, and after researching a little about whether Essentia is still _a thing_ and in active development, I figured that a useful solution on the long-run could be to support/invent actual 'essentia' file tags.

Now that brings us right back to the start I guess, but what do you think? And can you think of a "namespace preview" for the tagnames already. I was thinking something like 'ESSENTIA:' or 'ESS:', never liked the colons in the tagnames actually but no better idea currtently, just brainstorming.

JOJ0 commented:
I mean it doesn't look like it's dying tomorrow but you can never know actually...https://github.com/MTG/essentia/pulls
The maintainer I seems to be a member of the Barcelona university https://www.upf.edu/web/mtg/ that's invented Essentia.

JOJ0 commented:
And on another note: I am well aware of that no matter how these tags are called, they will always stay a niche and probably won't be very useable to a regular user because no software will really support it. Coming from the DJ side, I can see that the only way of making use of essentia gathered information is to 1.) translate it into a human readable form and 2.) bringing it into more widely used tags like `mood` or even `comment`. I always had that in mind but for that idea to realize at least with beets it's not needed to have extra tags for the "bare" essentia information.

Just saying to give some context on my thoughts overall on that topic and personal goals.

sampsyo commented:
Thanks for all the details! Well, maybe a good place to start would be asking the people involved with Essentia if they have considered standardizing metadata for the fields they can extract (presumably not using an AcousticBrainz prefix but some other namespacing?). That could inform whether it makes sense to add those directly to the `xtractor` plugin and call it a day?

JOJ0 commented:
Hi @sampsyo, lately I didn't find time to move this forward but in the meantime there's quite some new. First of have a look at my latest commit. I moved all the tag mappings from mediafile repo to the plugin directly. I changed the prefix to something that can live one once AcousticBrainz vanishes completely (They said read access until early 2023 in their blog post):

```
TXXX=Essentia_ab Average Loudness=0.887623310089
TXXX=Essentia_ab Chords Changes Rate=0.082755811512
TXXX=Essentia_ab Chords Key=A#
TXXX=Essentia_ab Chords Number Rate=0.002038320526
TXXX=Essentia_ab Chords Scale=minor
TXXX=Essentia_ab Danceable=0.000000000000
TXXX=Essentia_ab Gender=female
TXXX=Essentia_ab Genre Rosamerica=rhy
TXXX=Essentia_ab Ismir04 Rhythm=ChaChaCha
TXXX=Essentia_ab Key Strength=0.654201686382
TXXX=Essentia_ab Mood Acoustic=0.081802986562
TXXX=Essentia_ab Mood Aggressive=0.000000000000
TXXX=Essentia_ab Mood Electronic=0.979390621185
TXXX=Essentia_ab Mood Happy=0.085078120232
TXXX=Essentia_ab Mood Party=0.000015778420
TXXX=Essentia_ab Mood Relaxed=0.808817088604
TXXX=Essentia_ab Mood Sad=0.109234951437
TXXX=Essentia_ab Moods Mirex=Cluster5
TXXX=Essentia_ab Timbre=dark
TXXX=Essentia_ab Tonal=0.002889123978
TXXX=Essentia_ab Voice Instrumental=instrumental
```

When you look at the latest commit you see that I also tried to find versions of these tag keys more closely to what seems to be the standard way of tagging those file types.

I got in contact with @adamjakab (xtractor beets plugin) and @dbogdanov (MTG/Essentia), who helped me come to this proposal. Certainly the opinion of you two guys is very welcome for feedback on my proposal.

I tried to simplify the tags Picard used because I didn't like how complex and long the where (See example a couple of posts above)

Enough for now, y'all give me your thoughts and I hope we can find a solution that can be used in future plans regarding Essentia.

sampsyo commented:
Wow; that's awesome! Thanks for involving the folks from AB and Essentia in the conversation about picking tags. I like @dbogdanov's suggestion to use the longer `essentia:` prefix for long-term usefulness, and adding an additional `ab:` prefix for AB-specific stuff. Great teamwork! :smiley:

adamjakab commented:
Yes, truly. And I have no merit in this. I'm not even sure I remember how to write python code anymore.
But I have a quick comment: In the XTractor plugin I purposefully made two additional keys (`is_male` and `is_female`) for the `gender` field because it could be important to know how "strong" that prediction was. The same goes for `voice_instrumental`. Keeping the `is_voice` and `is_instrumental` values does not only provide numerical data (which is in many cases also easier to work with) but it keeps the detail which is lost when the categorization is made. Just a thought.
I also like the `Essentia_ab` prefix.

adamjakab commented:
> Thanks for all the details! Well, maybe a good place to start would be asking the people involved with Essentia if they have considered standardizing metadata for the fields they can extract (presumably not using an AcousticBrainz prefix but some other namespacing?). That could inform whether it makes sense to add those directly to the `xtractor` plugin and call it a day?

Sorry I am reading this thread backwards and I just read this now. 
Is your suggestion @sampsyo that these media fields should be handled by the `xtractor` plugin (given that the values are generated there)? From a software architecture point of view it would definitely make sense.

JOJ0 commented:
> Wow; that's awesome! Thanks for involving the folks from AB and Essentia in the conversation about picking tags. I like @dbogdanov's suggestion to use the longer `essentia:` prefix for long-term usefulness, and adding an additional `ab:` prefix for AB-specific stuff. Great teamwork! :smiley:

Glad you like it, but please clarify:

You dont like the Essentia_ab prefix? Prefer colon as delimiter instead of spaces?

Also with longer you mean keeping the stuff in-between like 'hi' and 'tonal' in eg 'chords_key'

Thanks

JOJ0 commented:
> Yes, truly. And I have no merit in this. I'm not even sure I remember how to write python code anymore.
> 
> But I have a quick comment: In the XTractor plugin I purposefully made two additional keys (`is_male` and `is_female`) for the `gender` field because it could be important to know how "strong" that prediction was. The same goes for `voice_instrumental`. Keeping the `is_voice` and `is_instrumental` values does not only provide numerical data (which is in many cases also easier to work with) but it keeps the detail which is lost when the categorization is made. Just a thought.

Yes the AB plugin as some shortcomings imho compared to your plugin. I like the is_fields actually. Not at computer right now but I recall Picard saves eg female and also not_female numerical values. AB provides it via an api endpoint, as well as a textform value which beets plug uses. this tells what endpoints the plug uses: https://github.com/beetbox/beets/blob/bf6689519170d7183448be9b5b3897ab35a826e2/beetsplug/acousticbrainz.py#L27

only the "positive" values are used and eg unused is not_danceable

> 
> I also like the `Essentia_ab` prefix.

JOJ0 commented:
> > Thanks for all the details! Well, maybe a good place to start would be asking the people involved with Essentia if they have considered standardizing metadata for the fields they can extract (presumably not using an AcousticBrainz prefix but some other namespacing?). That could inform whether it makes sense to add those directly to the `xtractor` plugin and call it a day?
> 
> Sorry I am reading this thread backwards and I just read this now. Is your suggestion @sampsyo that these media fields should be handled by the `xtractor` plugin (given that the values are generated there)? From a software architecture point of view it would definitely make sense.

My first approach was to put all the fields required to tag files with the acousticbrainz beets plugin into the mediafile library. Later and after some valuable discussion and advice from @sampsyo I reconsidered and now think that the best approach is to put fields into the plugins directly.

Both ways have their benefits but as you say from a architecture point of view it sounds right.

The beets acousticbrainz plugin could have a subset of the available field names we agree on but other plugins, eg. xtractor or even a new one @dbogdanov _might_ write could use this subset but could certainly add more fields as the development progresses. As long as we try to write down these names somewhere (Essentia website - Tagging suggestion?). For starters the documentation for the field names lies in the code and each one of us could follow that. But if the goal should be that others using Essentia stuff follow these names it would help to put it to a place a developer would look for it.

JOJ0 commented:
Almost done I thought but now having a hard time with the test. It tells me that MediaFields are already existing. It turns out that when I run the tests in test_acousticbrainz solely, they run perfectly fine:

```
(beets) jojo@gin ~/git/beets (acousticbrainz_write_tags) $ python -m unittest test.test_acousticbrainz.MapDataToSchemeTest.test_realistic
.
----------------------------------------------------------------------
Ran 1 test in 0.040s

OK
(beets) jojo@gin ~/git/beets (acousticbrainz_write_tags) $ python -m unittest test.test_acousticbrainz.MapDataToSchemeTest.test_composite
.
----------------------------------------------------------------------
Ran 1 test in 0.036s

OK
(beets) jojo@gin ~/git/beets (acousticbrainz_write_tags) $
```



But when running with the suite, it seems that mediafile "remembers" the fields that where added with the first test that was run and doesn't like if add_media_field() is run again:

```
======================================================================
ERROR: test_composite (test.test_acousticbrainz.MapDataToSchemeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jojo/git/beets/test/test_acousticbrainz.py", line 63, in test_composite
    ab = AcousticPlugin()
  File "/Users/jojo/git/beets/beetsplug/acousticbrainz.py", line 154, in __init__
    self.add_media_field('danceable', MediaField(
  File "/Users/jojo/git/beets/beets/plugins.py", line 204, in add_media_field
    mediafile.MediaFile.add_field(name, descriptor)
  File "/Users/jojo/git/mediafile/mediafile.py", line 1744, in add_field
    raise ValueError(
ValueError: property "danceable" already exists on MediaFile

======================================================================
ERROR: test_realistic (test.test_acousticbrainz.MapDataToSchemeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jojo/git/beets/test/test_acousticbrainz.py", line 70, in test_realistic
    ab = AcousticPlugin()
  File "/Users/jojo/git/beets/beetsplug/acousticbrainz.py", line 154, in __init__
    self.add_media_field('danceable', MediaField(
  File "/Users/jojo/git/beets/beets/plugins.py", line 204, in add_media_field
    mediafile.MediaFile.add_field(name, descriptor)
  File "/Users/jojo/git/mediafile/mediafile.py", line 1744, in add_field
    raise ValueError(
ValueError: property "danceable" already exists on MediaFile

======================================================================
ERROR: test_recurse (test.test_acousticbrainz.MapDataToSchemeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/jojo/git/beets/test/test_acousticbrainz.py", line 38, in test_recurse
    ab = AcousticPlugin()
  File "/Users/jojo/git/beets/beetsplug/acousticbrainz.py", line 154, in __init__
    self.add_media_field('danceable', MediaField(
  File "/Users/jojo/git/beets/beets/plugins.py", line 204, in add_media_field
    mediafile.MediaFile.add_field(name, descriptor)
  File "/Users/jojo/git/mediafile/mediafile.py", line 1744, in add_field
    raise ValueError(
ValueError: property "danceable" already exists on MediaFile

----------------------------------------------------------------------
Ran 4 tests in 0.049s

FAILED (errors=3)
```

JOJ0 commented:
Ok the reason is obvious. Each test inits its own version of AcousticPlugin(), which calls add_media_fields() in __init__ each time. This should be handled using setUp and tearDown functionality of unittest I think. Also I see a lot of things in other plugins tests that tell me that this plugin's tests are from a time before all that existed. Will try to get it working but it might be something obvious, any help appreciated.

sampsyo commented:
You're right; this kind of test fixture is not obvious at all! It's quite tricky stuff, unfortunately.

`unload_plugins` and stuff is not a bad idea by itself, but unfortunately that doesn't address the MediaFile field additions. 😕 I believe the only test we have that contends with modifying MediaFile is `test_plugin_mediafield`:
https://github.com/beetbox/beets/blob/3f6da68af2444b77807692023abfd607f0146750/test/test_plugin_mediafield.py#L46-L60

The salient thing here is that this test appears to do its _own_ special cleanup in the `finally:` clause there. Unfortunately, we probably need to do the same thing here, or else create our own utilities just for removing newly-added fields at the end of the test.

JOJ0 commented:
Hi @sampsyo, have a look, this is how I solved it. Thanks for pointing me in the right direction and most of all delattr()! 
My experiments were not too far of then, silly me tried to use del() instead :-o

Does it make any sense to also subclass like it's done in other tests. Using these modules:
```
from test import _common
from test.helper import TestHelper
```
and to use load/unload plugins?

Just for "standardizations" sake?

If not, I think we are good to go for a final review overall and hopefully ready to merge soon :-)

JOJ0 commented:
I accidently had commited a changed file in test/rsrc/, I kicked out this change and force pushed. Also I had missed a flake 8 "too many empty lines" error. Also fixed and now I think really really good to go :-) Forget about they ideas for "standardization" like in other tests, it's probably not worth the hassle. 

All green now, final-review-ready :-)

sampsyo commented:
I suppose that makes sense! Sorry for the bad timing with the AB shutdown. ☹️
